core/consensus/common/src/import_queue.rs

```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
+	sender: Sender<BlockImportMsg<B>>,
 }
 
-/// Locks order: queue, queue_blocks, best_importing_number
-pub struct AsyncImportQueueData<B: BlockT> {
-	signal: Condvar,
-	queue: Mutex<VecDeque<(BlockOrigin, Vec<IncomingBlock<B>>)>>,
-	queue_blocks: RwLock<HashSet<B::Hash>>,
-	best_importing_number: RwLock<<<B as BlockT>::Header as HeaderT>::Number>,
-	is_stopping: AtomicBool,
-}
-
-impl<B: BlockT, V: Verifier<B>> BasicQueue<B, V> {
-	/// Instantiate a new basic queue, with given verifier and justification import.
-	pub fn new(verifier: Arc<V>, block_import: SharedBlockImport<B>, justification_import: Option<SharedJustificationImport<B>>) -> Self {
-		Self {
-			handle: Mutex::new(None),
-			data: Arc::new(AsyncImportQueueData::new()),
-			verifier,
-			block_import,
-			justification_import,
-		}
+impl<B: BlockT> ImportQueueClone<B> for BasicQueue<B> {
+	fn clone_box(&self) -> Box<ImportQueue<B>> {
+		Box::new(self.clone())
 	}
 }
 
-impl<B: BlockT> AsyncImportQueueData<B> {
-	/// Instantiate a new async import queue data.
-	pub fn new() -> Self {
+impl<B: BlockT> BasicQueue<B> {
+	/// Instantiate a new basic queue, with given verifier.
+	pub fn new<V: 'static + Verifier<B>>(
+		verifier: Arc<V>,
+		block_import: SharedBlockImport<B>,
+		justification_import: Option<SharedJustificationImport<B>>
+	) -> Self {
+		let (result_sender, result_port) = channel::unbounded();
+		let worker_sender = BlockImportWorker::new(result_sender, verifier, block_import);
+		let importer_sender = BlockImporter::new(result_port, worker_sender, justification_import);
+
 		Self {
-			signal: Default::default(),
-			queue: Mutex::new(VecDeque::new()),
-			queue_blocks: RwLock::new(HashSet::new()),
-			best_importing_number: RwLock::new(Zero::zero()),
-			is_stopping: Default::default(),
+			sender: importer_sender,
 		}
 	}
-
-	// Signals to stop importing new blocks.
-	pub fn stop(&self) {
-		self.is_stopping.store(true, Ordering::SeqCst);
-	}
 }
 
-impl<B: BlockT, V: 'static + Verifier<B>> ImportQueue<B> for BasicQueue<B, V> {
-	fn start<L: 'static + Link<B>>(
-		&self,
-		link: L,
-	) -> Result<(), std::io::Error> {
-		debug_assert!(self.handle.lock().is_none());
-
-		let qdata = self.data.clone();
-		let verifier = self.verifier.clone();
-		let block_import = self.block_import.clone();
-		let justification_import = self.justification_import.clone();
-		*self.handle.lock() = Some(::std::thread::Builder::new().name("ImportQueue".into()).spawn(move || {
-			if let Some(justification_import) = justification_import.as_ref() {
-				justification_import.on_start(&link);
-			}
-			import_thread(block_import, link, qdata, verifier)
-		})?);
+impl<B: BlockT> ImportQueue<B> for BasicQueue<B> {
+	fn start(&self, link: Box<Link<B>>) -> Result<(), std::io::Error> {
+		let _ = self
+			.sender
+			.send(BlockImportMsg::Start(link))
+			.expect("1. self is holding a sender to the Importer, 2. Importer should handle messages while there are senders around");
```

Comment:
We usually write these proofs in the form:
"self is holding a sender to the Importer; Importer should handle messages while there are senders around; qed"